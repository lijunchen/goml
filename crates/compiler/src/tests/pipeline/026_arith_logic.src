fn show_int(label: string, value: int) {
    let _ = string_println(string_add(label, int_to_string(value))) in
    ()
}

fn show_bool(label: string, value: bool) {
    let _ = string_println(string_add(label, bool_to_string(value))) in
    ()
}

fn main() {
    let base = 10 in
    let sum = base + 5 in
    let diff = sum - 3 in
    let prod = diff * 2 in
    let quot = prod / 4 in

    let _ = show_int("sum=", sum) in
    let _ = show_int("diff=", diff) in
    let _ = show_int("prod=", prod) in
    let _ = show_int("quot=", quot) in

    let and_result = true && false in
    let or_result = true || false in
    let not_result = !false in
    let mixed = !and_result
        && match sum + prod * base - prod / 2 {
            0 => false,
            _ => true,
        }
        || !match diff - quot + base - sum / 2 {
            0 => false,
            _ => true,
        } in

    let _ = show_bool("and=", and_result) in
    let _ = show_bool("or=", or_result) in
    let _ = show_bool("not=", not_result) in
    let _ = show_bool("mixed=", mixed) in

    ()
}
