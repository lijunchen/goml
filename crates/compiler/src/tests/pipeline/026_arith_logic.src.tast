fn show_int(label/0: string, value/1: int) -> unit {
  let _ : unit = (string_println : (string) -> unit)((string_add : (string, string) -> string)((label/0 : string), (int_to_string : (int) -> string)((value/1 : int)))) in
  ()
}

fn show_bool(label/2: string, value/3: bool) -> unit {
  let _ : unit = (string_println : (string) -> unit)((string_add : (string, string) -> string)((label/2 : string), (bool_to_string : (bool) -> string)((value/3 : bool)))) in
  ()
}

fn main() -> unit {
  let base/4: int = 10 in
  let sum/5: int = (base/4 : int) + 5 in
  let diff/6: int = (sum/5 : int) - 3 in
  let prod/7: int = (diff/6 : int) * 2 in
  let quot/8: int = (prod/7 : int) / 4 in
  let _ : unit = (show_int : (string, int) -> unit)("sum=", (sum/5 : int)) in
  let _ : unit = (show_int : (string, int) -> unit)("diff=", (diff/6 : int)) in
  let _ : unit = (show_int : (string, int) -> unit)("prod=", (prod/7 : int)) in
  let _ : unit = (show_int : (string, int) -> unit)("quot=", (quot/8 : int)) in
  let and_result/9: bool = true && false in
  let or_result/10: bool = true || false in
  let not_result/11: bool = !false in
  let mixed/12: bool = !(and_result/9 : bool) && match (sum/5 : int) + (prod/7 : int) * (base/4 : int) - (prod/7 : int) / 2 {
      0 => false,
      _ : int => true,
  } || !match (diff/6 : int) - (quot/8 : int) + (base/4 : int) - (sum/5 : int) / 2 {
      0 => false,
      _ : int => true,
  } in
  let _ : unit = (show_bool : (string, bool) -> unit)("and=", (and_result/9 : bool)) in
  let _ : unit = (show_bool : (string, bool) -> unit)("or=", (or_result/10 : bool)) in
  let _ : unit = (show_bool : (string, bool) -> unit)("not=", (not_result/11 : bool)) in
  let _ : unit = (show_bool : (string, bool) -> unit)("mixed=", (mixed/12 : bool)) in
  ()
}