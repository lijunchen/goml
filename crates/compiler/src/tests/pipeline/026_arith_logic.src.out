sum=15
diff=12
prod=24
quot=6
and=false
or=true
not=true
mixed=true
